gh-111138: Add PyList_Extend() and PyList_Clear() functions (#111862)
python/cpython: Doc/c-api/list.rst

   ``len(list)`` on a list object.

.. c:function:: int PyList_Clear(PyObject *list)

   Remove all items from *list*. Similar to:

>    ``PyList_SetSlice(L, 0, PY_SSIZE_T_MAX, NULL)``.

Reviewer: ```suggestion
   ``PyList_SetSlice(list, 0, PY_SSIZE_T_MAX, NULL)``.
```
What does it mean to be similar? Does it have exactly the same effect? If not, what is the difference

Author: `PyList_Clear(list)` is the same as `PyList_SetSlice(list, 0, PY_SSIZE_T_MAX, NULL)`. Internally, both end up calling `list_clear()`: same code path.

>    Remove all items from *list*. Similar to:
>    ``PyList_SetSlice(list, 0, PY_SSIZE_T_MAX, NULL)``.

Reviewer 2: ```suggestion
   Remove all items from *list*.
   This is the same as ``PyList_SetSlice(list, 0, PY_SSIZE_T_MAX, NULL)`` and
   analogous to ``list.clear()`` or ``del list[:]``.
```

   ``len(list)`` on a list object.

.. c:function:: int PyList_Clear(PyObject *list)

   Remove all items from *list*. Similar to:

>    ``PyList_SetSlice(list, 0, PY_SSIZE_T_MAX, NULL)``.

Reviewer 3: Should we way anything about whether the underlying storage is resized?

Author: https://docs.python.org/dev/c-api/list.html documentation doesn't say anything about memory allocation, overallocation or freeing the internal storage. I would prefer to not start to document internals, unless there is a good reasons to document the exact behavior of PyList_Clear() related to memory handling.

   Raise an exception and return ``-1`` if *list* is not a :class:`list`
   object.  Return 0 on success.

   .. note::

>       If the finalizer (:meth:`object.__del__`) of a deleted item modify the

Reviewer 3: Is it a useful note for the reader? This is a particularly unlikely situation, unless you're deliberately provoking it.

Author: If you ask the question, you consider that it's not useful, so I removed the note.

   to ``list.append(item)``.

.. c:function:: int PyList_Extend(PyObject *self, PyObject *iterable)

   Extend *list* with the contents of *iterable*. Similar to:

>    ``PyList_SetSlice(L, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, iterable)``.

Reviewer: ```suggestion
   ``PyList_SetSlice(list, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, iterable)``.
```

Author: Oh right, fixed.

Reviewer 2: It has the same effect except that they raise different exceptions if the first argument is not a list.

Why is this? Why not simply raise a SystemError like in most of other Concrete Objects Layer C API?

Author: I would like to ask the question differently: why do other functions raise SystemError instead of TypeError?

In Python, TypeError is raised:

```
>>> list.extend("abc", "def")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'extend' for 'list' objects doesn't apply to a 'str' object

>>> list.clear("abc")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'clear' for 'list' objects doesn't apply to a 'str' object
```

Reviewer 2: Because the simplest alternative to SystemError is no check at all. And an undefined behavior that often is manifested as a crash is the result. SystemError is a signal that the C API is used improperly. No correct program should get a SystemError.

1. At the low level: no checks, an undefined behavior, crash.
2. A higher Concrete Objects Layer: SystemError.
3. At Abstract Objects Layer and in Python: TypeError.

>    Extend *list* with the contents of *iterable*. Similar to:
>    ``PyList_SetSlice(list, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, iterable)``.

Reviewer 2: ```suggestion
   Extend *list* with the contents of *iterable*.
   This is the same as ``PyList_SetSlice(list, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, iterable)``
   and analogous to ``list.extend(iterable)`` or ``list += iterable``.
```

   Return ``0`` on success, ``-1`` on failure.  Indexing from the end of the
   list is not supported.

   .. note::

      If *itemlist* is ``NULL`` to delete items and the finalizer
      (:meth:`object.__del__`) of a deleted item modify the list, it is

>       possible that the list is longer than expected.

Reviewer 3: Same question: is it useful to the reader?

Author: If you ask the question, you consider that it's not useful, so I removed the note.
